Block.timestamp implementations incorrectly returns Fix64 instead of UFix64

### Problem

I was doing a forced conversion to my own NFT type and it crashed my flow emulator. 

_Disclaimer: I'm completely new so can't provide better feedback and do more research but wanted to report and can provide more info if needed._

### Steps to Reproduce
This is the line that causes the crash. Note the ! at the dictionary to convert from optional and after as to force conversion to my own NFT.

`let kitty <- self.ownedNFTs.remove(key: id)! as! @Kitty.NFT`

This is my Kitty.NFT (implementation code not relevant imho - so commented out):

```
pub resource NFT: NonFungibleToken.INFT {
        // unique id of kitty
        pub let id: UInt64
        
        // energy consumption for the kitty energy units per minutes
        pub let energyConsumptionPerMinute: UFix64

        // max energy kitty has when it's completely fed
        pub let maxEnergy: UFix64
        
        // current energy level kitty has
        pub var energy: UFix64 // change to UInt8 with food type
        
        // track when the kitty was last fed
        pub var lastFed: UFix64
        
        init(id: UInt64) { /** init code */  }

        pub fun feedMe(food: @FungibleToken.Vault) {  /**. code */  }

        pub fun updateEnergy() {  /** code */  }
    }   
```


### Context

This is the output I have:
```
panic: interface conversion: interpreter.NumberValue is interpreter.UFix64Value, not interpreter.Fix64Value [recovered]
        panic: interface conversion: interpreter.NumberValue is interpreter.UFix64Value, not interpreter.Fix64Value [recovered]
        panic: interface conversion: interpreter.NumberValue is interpreter.UFix64Value, not interpreter.Fix64Value

goroutine 1754 [running]:
github.com/onflow/cadence/runtime/interpreter.recoverErrors(0xc000835298)
        github.com/onflow/cadence@v0.12.3/runtime/interpreter/interpreter.go:895 +0x15a
panic(0x4d8dd80, 0xc000b84e70)
        runtime/panic.go:969 +0x1b9
github.com/onflow/cadence/runtime/interpreter.recoverErrors(0xc000835218)
        github.com/onflow/cadence@v0.12.3/runtime/interpreter/interpreter.go:895 +0x15a
panic(0x4d8dd80, 0xc000b84e70)
        runtime/panic.go:969 +0x1b9
github.com/onflow/cadence/runtime/interpreter.Fix64Value.Div(0x23ca645f0b76400, 0x520cd80, 0xc000c6d768, 0x203000, 0x8)
        github.com/onflow/cadence@v0.12.3/runtime/interpreter/value.go:4986 +0x1fb
github.com/onflow/cadence/runtime/interpreter.(*Interpreter).VisitBinaryExpression.func5(0x520c680, 0xc000c6d4e0, 0x520cd80, 0xc000c6d768, 0xc000c6d768, 0x203000)
        github.com/onflow/cadence@v0.12.3/runtime/interpreter/interpreter.go:1697 +0x45
github.com/onflow/cadence/runtime/interpreter.(*Interpreter).visitNumberBinaryOperation.func1(0x4dedfc0, 0xc0010d9560, 0x18, 0x4e0dae0)
        github.com/onflow/cadence@v0.12.3/runtime/interpreter/interpreter.go:1655 +0xe2
github.com/onflow/cadence/runtime/trampoline.MapTrampoline.func1(0x4dedfc0, 0xc0010d9560, 0x4e0db80, 0xc0010d9760)
        github.com/onflow/cadence@v0.12.3/runtime/trampoline/trampoline.go:70 +0x39
github.com/onflow/cadence/runtime/trampoline.FlatMap.FlatMap.func1(0x4dedfc0, 0xc0010d9560, 0xc0010d9780, 0x4e7c3c0)
        github.com/onflow/cadence@v0.12.3/runtime/trampoline/trampoline.go:142 +0x42
github.com/onflow/cadence/runtime/trampoline.FlatMap.FlatMap.func1(0x4dedfc0, 0xc0010d9560, 0xc0008350f0, 0xc0008350e0)
        github.com/onflow/cadence@v0.12.3/runtime/trampoline/trampoline.go:142 +0x42
github.com/onflow/cadence/runtime/trampoline.FlatMap.FlatMap.func1(0x4dedfc0, 0xc0010d9560, 0xc000835160, 0x44d4c65)
        github.com/onflow/cadence@v0.12.3/runtime/trampoline/trampoline.go:142 +0x42
github.com/onflow/cadence/runtime/trampoline.FlatMap.FlatMap.func1(0x4dedfc0, 0xc0010d9560, 0xc0010d8600, 0x1339a724051f3e40)
        github.com/onflow/cadence@v0.12.3/runtime/trampoline/trampoline.go:142 +0x42
github.com/onflow/cadence/runtime/trampoline.FlatMap.FlatMap.func1(0x4dedfc0, 0xc0010d9560, 0x51f3e40, 0xc0010d9560)
        github.com/onflow/cadence@v0.12.3/runtime/trampoline/trampoline.go:142 +0x42
github.com/onflow/cadence/runtime/trampoline.FlatMap.FlatMap.func1(0x4dedfc0, 0xc0010d9560, 0x4d03280, 0xc0010d9780)
        github.com/onflow/cadence@v0.12.3/runtime/trampoline/trampoline.go:142 +0x42
github.com/onflow/cadence/runtime/trampoline.FlatMap.Resume.func1(0xc0010d9760, 0x4d03280)
        github.com/onflow/cadence@v0.12.3/runtime/trampoline/trampoline.go:154 +0x37
github.com/onflow/cadence/runtime/interpreter.(*Interpreter).runUntilNextStatement(0xc00001e1c0, 0x51f3e40, 0xc0010d9760, 0x0, 0x0, 0xc000b84db0)
        github.com/onflow/cadence@v0.12.3/runtime/interpreter/interpreter.go:610 +0x32
github.com/onflow/cadence/runtime/interpreter.(*Interpreter).runAllStatements(0xc00001e1c0, 0x51f3e40, 0xc000123aa0, 0x0, 0x0)
        github.com/onflow/cadence@v0.12.3/runtime/interpreter/interpreter.go:651 +0xb0
github.com/onflow/cadence/runtime/interpreter.(*Interpreter).InvokeTransaction(0xc00001e1c0, 0x0, 0xc000e44c80, 0x3, 0x4, 0x0, 0x0)
        github.com/onflow/cadence@v0.12.3/runtime/interpreter/interpreter.go:884 +0x11d
github.com/onflow/cadence/runtime.(*interpreterRuntime).transactionExecutionFunction.func1(0xc00001e1c0, 0x0, 0x0, 0x2286499000835460, 0x60131be4)
        github.com/onflow/cadence@v0.12.3/runtime/runtime.go:445 +0x165
github.com/onflow/cadence/runtime.(*interpreterRuntime).interpret.func1()
        github.com/onflow/cadence@v0.12.3/runtime/runtime.go:283 +0xa4
github.com/onflow/cadence/runtime.reportMetric(0xc000835518, 0x520e120, 0xc0000ef520, 0xc000835540)
        github.com/onflow/cadence@v0.12.3/runtime/runtime.go:146 +0x93
github.com/onflow/cadence/runtime.(*interpreterRuntime).interpret(0xc0000113f0, 0x520e120, 0xc0000ef520, 0x51f3d00, 0xc000b6b280, 0x0, 0x0, 0x0, 0xc00079ed80, 0xc00079ee10, ...)
        github.com/onflow/cadence@v0.12.3/runtime/runtime.go:277 +0x255
github.com/onflow/cadence/runtime.(*interpreterRuntime).ExecuteTransaction(0xc0000113f0, 0xc000eaa340, 0x194, 0x1a0, 0xc000c126c0, 0x2, 0x2, 0x520e120, 0xc0000ef520, 0x51f3d00, ...)
        github.com/onflow/cadence@v0.12.3/runtime/runtime.go:387 +0x5cf
github.com/onflow/flow-go/fvm.(*TransactionInvocator).Process(0xc000426840, 0xc0000a4c00, 0x52017e0, 0x5b59f10, 0x51e0c00, 0xc0000113f8, 0x51d1d20, 0xc0000a4be0, 0x0, 0x186a0, ...)
        github.com/onflow/flow-go@v0.13.1-0.20210119024156-79985711a957/fvm/transaction.go:80 +0x2e3
github.com/onflow/flow-go/fvm.(*TransactionProcedure).Run(0xc000d1a680, 0xc0000a4c00, 0x52017e0, 0x5b59f10, 0x51e0c00, 0xc0000113f8, 0x51d1d20, 0xc0000a4be0, 0x0, 0x186a0, ...)
        github.com/onflow/flow-go@v0.13.1-0.20210119024156-79985711a957/fvm/transaction.go:41 +0xd7
github.com/onflow/flow-go/fvm.(*VirtualMachine).Run(0xc0000a4c00, 0x52017e0, 0x5b59f10, 0x51e0c00, 0xc0000113f8, 0x51d1d20, 0xc0000a4be0, 0x0, 0x186a0, 0x3e80, ...)
        github.com/onflow/flow-go@v0.13.1-0.20210119024156-79985711a957/fvm/fvm.go:36 +0x162
github.com/onflow/flow-emulator.(*Blockchain).executeNextTransaction.func1(0xc000c13350, 0xc000000001, 0xc00126ac00, 0xc001070528, 0x20, 0x0)
        github.com/onflow/flow-emulator@v0.14.0/blockchain.go:790 +0x1f2
github.com/onflow/flow-emulator.(*pendingBlock).ExecuteNextTransaction(0xc00011d950, 0xc00115e7e8, 0x7, 0x0, 0x0)
        github.com/onflow/flow-emulator@v0.14.0/pendingBlock.go:151 +0x12d
github.com/onflow/flow-emulator.(*Blockchain).executeNextTransaction(0xc00042e340, 0x52017e0, 0x5b59f10, 0x51e0c00, 0xc0000113f8, 0x51d1d20, 0xc0000a4be0, 0x0, 0x186a0, 0x3e80, ...)
        github.com/onflow/flow-emulator@v0.14.0/blockchain.go:782 +0x7d
github.com/onflow/flow-emulator.(*Blockchain).executeBlock(0xc00042e340, 0x4042e1c, 0xc000725e00, 0x300000002, 0xc000725e00, 0xc00115ee30)
        github.com/onflow/flow-emulator@v0.14.0/blockchain.go:746 +0x238
github.com/onflow/flow-emulator.(*Blockchain).executeAndCommitBlock(0xc00042e340, 0x406e336, 0x4073780, 0xc00115ee68, 0x0, 0xc00115eeb0, 0x40c44c6)
        github.com/onflow/flow-emulator@v0.14.0/blockchain.go:873 +0x2f
github.com/onflow/flow-emulator.(*Blockchain).ExecuteAndCommitBlock(0xc00042e340, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        github.com/onflow/flow-emulator@v0.14.0/blockchain.go:867 +0x95
github.com/onflow/flow-emulator/server/backend.(*Backend).CommitBlock(0xc0002ddfa0)
        github.com/onflow/flow-emulator@v0.14.0/server/backend/backend.go:524 +0x4c
github.com/onflow/flow-emulator/server/backend.(*Backend).SendTransaction(0xc0002ddfa0, 0x51f2200, 0xc000c12600, 0xc000eaa340, 0x194, 0x1a0, 0xc000c126c0, 0x2, 0x2, 0xb670ded013aac8e6, ...)
        github.com/onflow/flow-emulator@v0.14.0/server/backend/backend.go:237 +0x6c6
github.com/onflow/flow-emulator/server/backend.(*Adapter).SendTransaction(0xc0001b4698, 0x51f2200, 0xc000c12600, 0xc00126a300, 0x428423f7686b2c76, 0x71ed83eb744cd511)
        github.com/onflow/flow-emulator@v0.14.0/server/backend/adapter.go:82 +0xf8
github.com/onflow/flow-go/access/legacy.(*Handler).SendTransaction(0xc0002fc040, 0x51f2200, 0xc000c12600, 0xc000c12630, 0xc0002fc040, 0xe859f20, 0xc0003bf980)
        github.com/onflow/flow-go@v0.13.1-0.20210119024156-79985711a957/access/legacy/handler.go:52 +0x23c
github.com/onflow/flow/protobuf/go/flow/legacy/access._AccessAPI_SendTransaction_Handler.func1(0x51f2200, 0xc000c12600, 0x4e96b80, 0xc000c12630, 0x21, 0xc000678370, 0x0, 0xc001155b30)
        github.com/onflow/flow/protobuf/go/flow@v0.1.8/legacy/access/access.pb.go:1993 +0x8b
github.com/grpc-ecosystem/go-grpc-prometheus.(*ServerMetrics).UnaryServerInterceptor.func1(0x51f2200, 0xc000c12600, 0x4e96b80, 0xc000c12630, 0xc00038c200, 0xc00038c240, 0x4071a87, 0x411eb05, 0x4e80e00, 0xc000c12600)
        github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:107 +0xad
github.com/onflow/flow/protobuf/go/flow/legacy/access._AccessAPI_SendTransaction_Handler(0x4ef41a0, 0xc0002fc040, 0x51f2200, 0xc000c12600, 0xc0001f97a0, 0xc00019d290, 0x51f2200, 0xc000c12600, 0xc000f70a00, 0x27f)
        github.com/onflow/flow/protobuf/go/flow@v0.1.8/legacy/access/access.pb.go:1995 +0x150
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001c8e00, 0x5201900, 0xc00106e120, 0xc0011c8c00, 0xc000159920, 0x5b6c640, 0x0, 0x0, 0x0)
        google.golang.org/grpc@v1.32.0/server.go:1194 +0x522
google.golang.org/grpc.(*Server).handleStream(0xc0001c8e00, 0x5201900, 0xc00106e120, 0xc0011c8c00, 0x0)
        google.golang.org/grpc@v1.32.0/server.go:1517 +0xd05
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc00021d440, 0xc0001c8e00, 0x5201900, 0xc00106e120, 0xc0011c8c00)
        google.golang.org/grpc@v1.32.0/server.go:859 +0xa5
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.32.0/server.go:857 +0x1fd
```

--- a/runtime/interpreter/block.go
+++ b/runtime/interpreter/block.go
@@ -32,7 +32,7 @@ type BlockValue struct {
 	Height    UInt64Value
 	View      UInt64Value
 	ID        *ArrayValue
-	Timestamp Fix64Value
+	Timestamp UFix64Value
 }
 
 func (BlockValue) IsValue() {}

--- a/runtime/interpreter/value_test.go
+++ b/runtime/interpreter/value_test.go
@@ -626,3 +626,11 @@ func TestVisitor(t *testing.T) {
 	require.Equal(t, 1, intVisits)
 	require.Equal(t, 1, stringVisits)
 }
+
+func TestBlockValue(t *testing.T) {
+	var block BlockValue = BlockValue{ 4, 5, &ArrayValue{}, 5.0}
+	// static type test
+	var actualTs UFix64Value = block.Timestamp
+	const expectedTs UFix64Value = 5.0
+	assert.Equal(t, expectedTs, actualTs)
+}

--- a/runtime/pretty/print.go
+++ b/runtime/pretty/print.go
@@ -253,7 +253,7 @@ func (p ErrorPrettyPrinter) writeCodeExcerpts(
 		}
 
 		// code, if position
-		if excerpt.startPos != nil && len(code) > 0 {
+		if excerpt.startPos != nil && excerpt.startPos.Line > 0 && len(code) > 0 {
 
 			if i > 0 && lastLineNumber != 0 && excerpt.startPos.Line-1 > lastLineNumber {
 				p.writeCodeExcerptContinuation(lineNumberLength)

--- a/runtime/runtime.go
+++ b/runtime/runtime.go
@@ -1879,7 +1879,7 @@ func NewBlockValue(block Block) interpreter.BlockValue {
 
 	// timestamp
 	// TODO: verify
-	timestampValue := interpreter.NewFix64ValueWithInteger(time.Unix(0, block.Timestamp).Unix())
+	timestampValue := interpreter.NewUFix64ValueWithInteger(uint64(time.Unix(0, block.Timestamp).Unix()))
 
 	return interpreter.BlockValue{
 		Height:    heightValue,

--- a/runtime/tests/checker/function_test.go
+++ b/runtime/tests/checker/function_test.go
@@ -369,3 +369,19 @@ func TestCheckInvalidResourceCapturingJustMemberAccess(t *testing.T) {
 	assert.IsType(t, &sema.ResourceCapturingError{}, errs[0])
 	assert.IsType(t, &sema.ResourceLossError{}, errs[1])
 }
+
+func TestCheckInvalidFunctionWithResult(t *testing.T) {
+
+	t.Parallel()
+
+	_, err := ParseAndCheck(t, `
+     fun test(): Int {
+         let result = 0
+         return result
+     }
+   `)
+
+	errs := ExpectCheckerErrors(t, err, 1)
+
+	assert.IsType(t, &sema.RedeclarationError{}, errs[0])
+}

--- a/runtime/type_test.go
+++ b/runtime/type_test.go
@@ -24,6 +24,7 @@ import (
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
+	"github.com/onflow/cadence"
 	"github.com/onflow/cadence/runtime/common"
 )
 
@@ -152,7 +153,6 @@ func TestBlockTimestamp(t *testing.T) {
 			pub fun main(): [UFix64] {
 				let block = getCurrentBlock()
 				var ts: UFix64 = block.timestamp
-				log(ts.isInstance(Type<UFix64>()))
 
 				var div: UFix64 = 4.0
 
@@ -163,26 +163,24 @@ func TestBlockTimestamp(t *testing.T) {
 			}
         `)
 
-		var loggedMessage string
-
-		runtimeInterface := &testRuntimeInterface{
-			log: func(message string) {
-				loggedMessage = message
-			},
-		}
-
-		_, err := runtime.ExecuteScript(
+		value, err := runtime.ExecuteScript(
 			Script{
 				Source: script,
 			},
 			Context{
-				Interface: runtimeInterface,
+				Interface: &testRuntimeInterface{},
 				Location:  common.ScriptLocation{},
 			},
 		)
 
 		require.NoError(t, err)
-		assert.Equal(t, "true", loggedMessage, "Block.timestamp is not UFix64")
+		require.NotNil(t, value)
 
+		require.IsType(t, cadence.Array{}, value)
+		values := value.(cadence.Array).Values
+
+		require.Equal(t, 2, len(values))
+		assert.IsType(t, cadence.UFix64Type{}, values[0].Type())
+		assert.IsType(t, cadence.UFix64Type{}, values[1].Type())
 	})
 }

--- a/runtime/type_test.go
+++ b/runtime/type_test.go
@@ -91,13 +91,65 @@ func TestRuntimeTypeStorage(t *testing.T) {
 	assert.Equal(t, `"Int"`, loggedMessage)
 }
 
-
 func TestBlockTimestamp(t *testing.T) {
+
 	t.Parallel()
-	runtime := NewInterpreterRuntime()
-	script := []byte(`
-		transaction {
-			prepare() {
+
+	t.Run("transaction", func(t *testing.T) {
+
+		t.Parallel()
+
+		runtime := NewInterpreterRuntime()
+
+		script := []byte(`
+			transaction {
+				prepare() {
+					let block = getCurrentBlock()
+					var ts: UFix64 = block.timestamp
+					log(ts.isInstance(Type<UFix64>()))
+
+					var div: UFix64 = 4.0
+
+					// Shouldn't panic
+					var res = ts/div
+				}
+			}
+        `)
+
+		var loggedMessage string
+
+		runtimeInterface := &testRuntimeInterface{
+			getSigningAccounts: func() ([]Address, error) {
+				return nil, nil
+			},
+			log: func(message string) {
+				loggedMessage = message
+			},
+		}
+
+		nextTransactionLocation := newTransactionLocationGenerator()
+		err := runtime.ExecuteTransaction(
+			Script{
+				Source: script,
+			},
+			Context{
+				Interface: runtimeInterface,
+				Location:  nextTransactionLocation(),
+			},
+		)
+		require.NoError(t, err)
+		assert.Equal(t, "true", loggedMessage, "Block.timestamp is not UFix64")
+
+	})
+
+	t.Run("script", func(t *testing.T) {
+
+		t.Parallel()
+
+		runtime := NewInterpreterRuntime()
+
+		script := []byte(`
+			pub fun main(): [UFix64] {
 				let block = getCurrentBlock()
 				var ts: UFix64 = block.timestamp
 				log(ts.isInstance(Type<UFix64>()))
@@ -105,32 +157,32 @@ func TestBlockTimestamp(t *testing.T) {
 				var div: UFix64 = 4.0
 
 				// Shouldn't panic
-				var result = ts/div
+				var res = ts/div
+
+				return [ts, res]
 			}
+        `)
+
+		var loggedMessage string
+
+		runtimeInterface := &testRuntimeInterface{
+			log: func(message string) {
+				loggedMessage = message
+			},
 		}
-    `)
 
-	var loggedMessage string
+		_, err := runtime.ExecuteScript(
+			Script{
+				Source: script,
+			},
+			Context{
+				Interface: runtimeInterface,
+				Location:  common.ScriptLocation{},
+			},
+		)
 
-	runtimeInterface := &testRuntimeInterface{
-		getSigningAccounts: func() ([]Address, error) {
-			return nil, nil
-		},
-		log: func(message string) {
-			loggedMessage = message
-		},
-	}
+		require.NoError(t, err)
+		assert.Equal(t, "true", loggedMessage, "Block.timestamp is not UFix64")
 
-	nextTransactionLocation := newTransactionLocationGenerator()
-	err := runtime.ExecuteTransaction(
-		Script{
-			Source: script,
-		},
-		Context{
-			Interface: runtimeInterface,
-			Location:  nextTransactionLocation(),
-		},
-	)
-	require.NoError(t, err)
-	assert.Equal(t, "True", loggedMessage, "Block.Timestamp is not UFix64")
+	})
 }

--- a/runtime/type_test.go
+++ b/runtime/type_test.go
@@ -103,7 +103,6 @@ func TestBlockTimestamp(t *testing.T) {
 				log(ts.isInstance(Type<UFix64>()))
 
 				var div: UFix64 = 4.0
-				var x = ts as UFix64
 
 				// Shouldn't panic
 				var result = ts/div
@@ -133,5 +132,5 @@ func TestBlockTimestamp(t *testing.T) {
 		},
 	)
 	require.NoError(t, err)
-	assert.True(t, loggedMessage == "true", "Block.Timestamp is not UFix64")
+	assert.Equal(t, "True", loggedMessage, "Block.Timestamp is not UFix64")
 }

--- a/runtime/type_test.go
+++ b/runtime/type_test.go
@@ -90,3 +90,48 @@ func TestRuntimeTypeStorage(t *testing.T) {
 
 	assert.Equal(t, `"Int"`, loggedMessage)
 }
+
+
+func TestBlockTimestamp(t *testing.T) {
+	t.Parallel()
+	runtime := NewInterpreterRuntime()
+	script := []byte(`
+		transaction {
+			prepare() {
+				let block = getCurrentBlock()
+				var ts: UFix64 = block.timestamp
+				log(ts.isInstance(Type<UFix64>()))
+
+				var div: UFix64 = 4.0
+				var x = ts as UFix64
+
+				// Shouldn't panic
+				var result = ts/div
+			}
+		}
+    `)
+
+	var loggedMessage string
+
+	runtimeInterface := &testRuntimeInterface{
+		getSigningAccounts: func() ([]Address, error) {
+			return nil, nil
+		},
+		log: func(message string) {
+			loggedMessage = message
+		},
+	}
+
+	nextTransactionLocation := newTransactionLocationGenerator()
+	err := runtime.ExecuteTransaction(
+		Script{
+			Source: script,
+		},
+		Context{
+			Interface: runtimeInterface,
+			Location:  nextTransactionLocation(),
+		},
+	)
+	require.NoError(t, err)
+	assert.True(t, loggedMessage == "true", "Block.Timestamp is not UFix64")
+}

